wcfsetup/install/files/lib/acp/form/FirstTimeSetupLicenseForm.class.php

```diff
@@ -0,0 +1,199 @@
+<?php
+
+namespace wcf\acp\form;
+
+use CuyZ\Valinor\Mapper\MappingError;
+use CuyZ\Valinor\Mapper\Source\Source;
+use CuyZ\Valinor\MapperBuilder;
+use GuzzleHttp\Psr7\Request;
+use Psr\Http\Client\ClientExceptionInterface;
+use wcf\data\option\Option;
+use wcf\data\option\OptionAction;
+use wcf\data\package\update\server\PackageUpdateServerAction;
+use wcf\data\package\update\server\PackageUpdateServerList;
+use wcf\form\AbstractForm;
+use wcf\form\AbstractFormBuilderForm;
+use wcf\system\exception\PermissionDeniedException;
+use wcf\system\form\builder\container\FormContainer;
+use wcf\system\form\builder\field\CheckboxFormField;
+use wcf\system\form\builder\field\dependency\EmptyFormFieldDependency;
+use wcf\system\form\builder\field\TextFormField;
+use wcf\system\form\builder\field\validation\FormFieldValidationError;
+use wcf\system\form\builder\field\validation\FormFieldValidator;
+use wcf\system\form\builder\LanguageItemFormNode;
+use wcf\system\io\HttpFactory;
+use wcf\system\request\LinkHandler;
+use wcf\util\HeaderUtil;
+
+/**
+ * Sets up license data during first time setup.
+ *
+ * @author      Tim Duesterhus
+ * @copyright   2001-2023 WoltLab GmbH
+ * @license     GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ */
+final class FirstTimeSetupLicenseForm extends AbstractFormBuilderForm
+{
+    /**
+     * @inheritDoc
+     */
+    public $neededPermissions = ['admin.configuration.package.canEditServer'];
+
+    private array $apiResponse;
+
+    /**
+     * @inheritDoc
+     */
+    public function readParameters()
+    {
+        parent::readParameters();
+
+        if (\FIRST_TIME_SETUP_STATE == -1) {
+            throw new PermissionDeniedException();
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function createForm()
+    {
+        parent::createForm();
+
+        $this->form->appendChildren([
+            LanguageItemFormNode::create('explanation')
+                ->languageItem('wcf.acp.firstTimeSetup.license.explanation'),
+            $credentialsContainer = FormContainer::create('credentials')
+                ->label('wcf.acp.firstTimeSetup.license.credentials')
+                ->appendChildren([
+                    TextFormField::create('licenseNo')
+                        ->label('wcf.acp.package.update.licenseNo')
+                        ->required()
+                        ->maximumLength(255)
+                        ->placeholder('123456'),
+                    TextFormField::create('serialNo')
+                        ->label('wcf.acp.package.update.serialNo')
+                        ->required()
+                        ->maximumLength(255)
+                        ->placeholder('AB12-34CD-E56F-12AB-7890')
+                        ->addValidator(new FormFieldValidator('serialNo', function (TextFormField $serialNo) {
```

Comment:
This should possibly a separate private method. I’m really not a big fan of these huge `createForm()` methods that do not seem to end and require me to scroll four screens to the right.